x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/TopList.java

```diff
@@ -0,0 +1,181 @@
+
+    @FunctionInfo(
+        returnType = { "double", "integer", "long", "date" },
+        description = "Collects the top values for a field. Includes repeated values.",
+        isAggregation = true,
+        examples = @Example(file = "stats_top_list", tag = "top-list")
+    )
+    public TopList(
+        Source source,
+        @Param(
+            name = "field",
+            type = { "double", "integer", "long", "date" },
+            description = "The field to collect the top values for."
+        ) Expression field,
+        @Param(name = "limit", type = { "integer" }, description = "The maximum number of values to collect.") Expression limit,
+        @Param(
+            name = "order",
+            type = { "keyword" },
+            description = "The order to calculate the top values. Either `asc` or `desc`."
+        ) Expression order
+    ) {
+        super(source, field, Arrays.asList(limit, order));
+    }
+
+    public static TopList readFrom(PlanStreamInput in) throws IOException {
+        return new TopList(Source.readFrom(in), in.readExpression(), in.readExpression(), in.readExpression());
+    }
+
+    public void writeTo(PlanStreamOutput out) throws IOException {
+        source().writeTo(out);
+        List<Expression> fields = children();
+        assert fields.size() == 3;
+        out.writeExpression(fields.get(0));
+        out.writeExpression(fields.get(1));
+        out.writeExpression(fields.get(2));
+    }
+
+    private Expression limitField() {
+        return parameters().get(0);
+    }
+
+    private Expression orderField() {
+        return parameters().get(1);
+    }
+
+    private int limitValue() {
+        return (int) limitField().fold();
+    }
+
+    private String orderRawValue() {
+        return BytesRefs.toString(orderField().fold());
+    }
+
+    private boolean orderValue() {
+        return orderRawValue().equalsIgnoreCase(ORDER_ASC);
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        var typeResolution = isType(
+            field(),
+            dt -> dt == DataType.DATETIME || dt.isNumeric() && dt != DataType.UNSIGNED_LONG,
+            sourceText(),
+            FIRST,
+            "numeric except unsigned_long or counter types"
+        ).and(isFoldable(limitField(), sourceText(), SECOND))
+            .and(isType(limitField(), dt -> dt == DataType.INTEGER, sourceText(), SECOND, "integer"))
+            .and(isFoldable(orderField(), sourceText(), THIRD))
+            .and(isString(orderField(), sourceText(), THIRD));
+
+        if (typeResolution.unresolved()) {
+            return typeResolution;
+        }
+
+        var limit = limitValue();
+        var order = orderRawValue();
+
+        if (limit <= 0) {
+            return new TypeResolution(format(null, "Limit must be greater than 0 in [{}], found [{}]", sourceText(), limit));
```

Comment:
Can you add tests for this parameter restriction?

Reply:
As discussed, added an [issue here](https://github.com/elastic/elasticsearch/issues/109917) to add support for automatic aggregation testing, replicating that of scalar functions. They already support type resolution error testing, so it should fit well after having it.